### PR TITLE
Check for pip configuration when an externally managed environment is detected.

### PR DESCRIPTION
**Pull request by nuclearsandwich**
rosdep is designed to treat pip like an alternative system-level package manager. Deviating from this approach is not easily achievable without a significant rethinking of how pip packages are managed. In the meantime, we can at least instruct users how to restore the prior functionality.

Rather than inject the environment variable / config on behalf of the user, this change instructs them to make the necessary config changes themself, keeping them informed of the change their making to the system's new default.

I added a documentation page specific to this issue with additional configuration details and recipes for CI.

Seeks to reduce friction related to #978

### CODE REVIEW

**Comment by nuclearsandwich** [comment]
In addition to the test failures, this actually doesn't work as intended because `sudo` is stripping the environment. I'm hesitant to recommend the pip global config file as a default because the default location is variable. Alternatively, we could override the sudo_command for the pip installer to pass that environment variable through.

**Comment by ros-discourse** [comment]
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/rosdep-for-pip-is-broken-on-jazzy/38981/5

File: src/rosdep2/platforms/pip.py
Comment (christophebedard) [inline src/rosdep2/platforms/pip.py:50]:
```suggestion
rosdep installation of pip packages requires installing packages globally as root.
```

File: src/rosdep2/platforms/pip.py
Comment (christophebedard) [inline src/rosdep2/platforms/pip.py:97]:
```suggestion
        ].lower() in ("yes", "1", "true"):
```

**Comment by nuclearsandwich** [comment]
> In addition to the test failures, this actually doesn't work as intended because `sudo` is stripping the environment. I'm hesitant to recommend the pip global config file as a default because the default location is variable. Alternatively, we could override the sudo_command for the pip installer to pass that environment variable through.

As of https://github.com/ros-infrastructure/rosdep/pull/979/commits/2c76686afa24ff09f52642a4b72116c2051ae5e3 the installer is configured to pass `PIP_BREAK_SYSTEM_PACKAGES` through from the user environment. This works by default on each of my tested sudo environments (Archlinux, Ubuntu, Fedora) but a locked down sudo that disallows environment preservation or has an explicit allowlist will still create problems. My hope was to detect that the check passed but the install failed due to the externally managed target anyway and give a hint about sudo to suggest an alternative configuration method. But without changes, the pip installer never gets control back from `install_resolved()` so we'd have to add pip-specific hooks elsewhere in the code and I'm hesitant. I think I'm content to leave it here but with more robust documentation.

File: src/rosdep2/platforms/pip.py
Comment (nuclearsandwich) [inline src/rosdep2/platforms/pip.py:57]: Note that this documentation won't actually be live until the PR is merged and docs rebuild the next day. This will only affect those using rosdep from mainline or this branch. I'll wait until the docs rebuild to make a new rosdep release with this change.

File: doc/pip_and_pep_668.rst
Comment (christophebedard) [inline doc/pip_and_pep_668.rst:22]: it looks like this sentence was cut off.

File: doc/pip_and_pep_668.rst
Comment (christophebedard) [inline doc/pip_and_pep_668.rst:28]:
```suggestion
`Pip configuration files <pip-configuration>`_ can be used to set the desired behavior.
```

File: doc/pip_and_pep_668.rst
Comment (christophebedard) [inline doc/pip_and_pep_668.rst:29]:
```suggestion
Pip checks for global configuration files in ``XDG_CONFIG_DIRS``, as well as ``/etc/pip.conf``.
```

File: doc/pip_and_pep_668.rst
Comment (christophebedard) [inline doc/pip_and_pep_668.rst:6]: this document is a bit inconsistent with rosdep and `rosdep`, but I'll let you decide if you want to change anything.

File: src/rosdep2/platforms/pip.py
Comment (christophebedard) [inline src/rosdep2/platforms/pip.py:90]:
```suggestion
    managed" environments such as operating systems with included package
```

File: src/rosdep2/platforms/pip.py
Comment (christophebedard) [inline src/rosdep2/platforms/pip.py:99]: minor suggestion, but, if you want, you could improve readability a little bit by changing it to something like:

```suggestion
        if (
            'PIP_BREAK_SYSTEM_PACKAGES' in os.environ
            and os.environ['PIP_BREAK_SYSTEM_PACKAGES'].lower() in ('yes', '1', 'true')
        ):
```

not actually sure if exact change passes linters, though.

File: src/rosdep2/platforms/pip.py
Comment (christophebedard) [inline src/rosdep2/platforms/pip.py:96]: you could flip this to return early and remove one indent level from the rest of the function

```suggestion
    if sys.version_info < (3, 11):
        return True
```

File: test/test_rosdep_pip.py
Comment (christophebedard) [inline test/test_rosdep_pip.py:70]: would be nice to test that the command fails if this isn't set and the version is >= 3.11, but you'd also need to make sure that the other config methods are not used, or just patch `externally_managed_installable()`, but I would leave that up to you.

**Review by christophebedard — COMMENTED** [review]
Minor comments!

File: doc/pip_and_pep_668.rst
Comment (nuclearsandwich) [reply doc/pip_and_pep_668.rst:6]: My mental style guide is `rosdep` when referring to the CLI tool and rosdep when referring to the project. I am always tempted into using `rosdep` in appropriately when its the first word of the sentence because Rosdep and ROSdep are both wrong.

I will reword to avoid the entire problem :D (that mainly exists in my head)

Comment (nuclearsandwich) [reply doc/pip_and_pep_668.rst:6]: Actually, I'm just going to get over myself and start the sentence with the proper lowercase noun. It's the 2020s.

File: doc/pip_and_pep_668.rst
Comment (nuclearsandwich) [reply doc/pip_and_pep_668.rst:22]: lol good catch. I'm sure that I went to look up how to reference a section in the doc below and got distracted because that's exactly what I just did again.

File: src/rosdep2/platforms/pip.py
Comment (nuclearsandwich) [reply src/rosdep2/platforms/pip.py]: The only think flake8 got upset by was the and being on the continuing line rather than the end of the first. I like this form better thanks.

File: test/test_rosdep_pip.py
Comment (nuclearsandwich) [reply test/test_rosdep_pip.py:86]: I'm limited somewhat by my lack of experience with Python's mocking. Let me see if I can skim enough examples and read enough of the docs to stub the method for a few tests before.

It's a good push. Tests added in 8b48f13

File: test/test_rosdep_pip.py
Comment (nuclearsandwich) [inline test/test_rosdep_pip.py:73]: I'm not super up on why imports are done per test function here but I am propagating the convention until I understand it better.

File: src/rosdep2/platforms/pip.py
Comment (nuclearsandwich) [reply src/rosdep2/platforms/pip.py]: This suggestion was incorporated into a2e9a1e

Comment (christophebedard) [reply src/rosdep2/platforms/pip.py]: I think you have an extra indentation level there:

```diff
     if (
-             'PIP_BREAK_SYSTEM_PACKAGES' in os.environ and
-             os.environ['PIP_BREAK_SYSTEM_PACKAGES'].lower() in ('yes', '1', 'true')
+         'PIP_BREAK_SYSTEM_PACKAGES' in os.environ and
+         os.environ['PIP_BREAK_SYSTEM_PACKAGES'].lower() in ('yes', '1', 'true')
     ):
         return True
```

in my experience the usual linters don't complain about this for some reason.

> The only think flake8 got upset by was the and being on the continuing line rather than the end of the first.

Ah yeah I always get bit by that!

File: test/test_rosdep_pip.py
Comment (christophebedard) [inline test/test_rosdep_pip.py:81]: If you want, a more Pythonic way could be:

```suggestion
    with pytest.raises(InstallFailed) as exception_info:
        installer.get_install_command(['whatever'])
    assert exception_info.value.failures == [('pip', EXTERNALLY_MANAGED_EXPLAINER)]
```

and make sure to `import pytest`

**Review by christophebedard — APPROVED** [review]
Aside from my last minor suggestions, this looks good to me!

File: src/rosdep2/platforms/pip.py
Comment (christophebedard) [inline src/rosdep2/platforms/pip.py:119]: I gave it a try and it seems like, if:

1. the config file doesn't exist; or
2. the `[install]` section doesn't exist; or
3. a `break-system-package` value isn't specified.

then `pip_config[...]` raises a `KeyError`. Looks like you could avoid checking all of that by using `getboolean()`:

```py
pip_config.getboolean('install', 'break-system-packages', fallback=False)
```

File: test/test_rosdep_pip.py
Comment (nuclearsandwich) [reply test/test_rosdep_pip.py:81]: Thanks for the suggestion. I'm actually going to keep this as is for now, because I don't believe the infrastructure packages ever explicitly require pytest. I think pytest is supposed to be compatible with stdlib unittest as well and utilizing that may be the preferred approach.

File: src/rosdep2/platforms/pip.py
Comment (nuclearsandwich) [reply src/rosdep2/platforms/pip.py:119]: Good catch!

Comment (nuclearsandwich) [reply src/rosdep2/platforms/pip.py]: Resolved in [8e94ef1](https://github.com/ros-infrastructure/rosdep/pull/979/commits/8e94ef1ffb01394b23cea9dcc12fbb6fc92f10d2)

**Review by christophebedard — APPROVED** [review]

**Pull request merged**
